### ISSUE TITLE
mdates.ConciseDateFormatter() doesn't work with zero_formats parameter 

### ISSUE THREAD

**Issue by fifzteen** [matplotlib/matplotlib#15178]
### Bug report

**Bug summary**

mdates.ConciseDateFormatter() raises TypeError when giving zero_formats parameter.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import numpy as np

ny = 2
xmin, xmax = mdates.datestr2num(['2019-09-04 00:00:00', '2019-09-05 00:00:00'])
data = np.random.random((ny, int(xmax-xmin)+1))

fig, ax = plt.subplots()
ax.imshow(data, extent=[xmin, xmax, 0, ny], aspect='auto')
xlocator = mdates.AutoDateLocator()
ax.xaxis.set_major_locator(xlocator)
ax.xaxis.set_major_formatter(mdates.ConciseDateFormatter(xlocator, zero_formats=['', '%Y', '%m', '%m-%d', '%H:%M', '%H:%M']))
plt.show()
```

**Actual outcome**

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\foobar\AppData\Local\Programs\Python\Python37\lib\site-packages\matplotlib\dates.py", line 842, in __init__
    if len(formats) != 6:
TypeError: object of type 'NoneType' has no len()
```

**Expected**

dates.py around line 842 below.

```python
841 if zero_formats:
842    if len(formats) != 6:
843        raise ValueError('zero_formats argument must be a list of '
844                                  '6 format strings (or None)')
845    self.zero_formats = zero_formats
```

now, formats is None (dafeult value). so, line 842 shuold be

```python
842    if len(zero_formats) != 6:
```

**Matplotlib version**

  * Operating system: Windows 10
  * Matplotlib version: 3.1.1
  * Python version: 3.7.4

**Comment by jklymak** [comment]
Hah I just copied that code over.  I had no idea what zero format is.  If you think it’s useful to make work with the formatter we should fix it, or if not, remove it.  Feel free to make a PR if you think you understand the issue well enough, otherwise I’ll try to fix in
A few weeks.  Thanks for the report

**Comment by fifzteen** [comment]
Thanks for your reply.

I simplify abobe example code to look the problem clearly.
What I want to get is a plot like below.
![Figure_1](https://user-images.githubusercontent.com/25410805/64231463-ae381d80-cf2a-11e9-880c-8ceded1333e6.png)

"zero_formats" is useful to draw date and time only the time is 0:00. like this code.

```python
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import numpy as np

ny = 2
xmin, xmax = mdates.datestr2num(['2019-09-04 00:00:00', '2019-09-05 00:00:00'])
data = np.random.random((ny, int(xmax-xmin)+1))

fig, ax = plt.subplots()
ax.imshow(data, extent=[xmin, xmax, 0, ny], aspect='auto')
xlocator = mdates.AutoDateLocator()
ax.xaxis.set_major_locator(xlocator)
ax.xaxis.set_major_formatter(mdates.ConciseDateFormatter(xlocator,
                                                         formats=['%Y', '%b', '%d', '%H:%M', '%H:%M', '%S.%f'],
                                                         zero_formats=['', '%Y', '%m', '%m-%d\n%H:%M', '%H:%M', '%H:%M'],
                                                         show_offset=False))
ax.tick_params(axis='x', labelrotation=90.)
plt.show()
```

(We can prevent this problem by setting formats to default value.)

I don't know how many effects to other codes are exists. so I think it's better that you fix this bug.

**Comment by jklymak** [comment]
Yeah but that’s what concisedateformatter is supposed to do automatically so I’m not sure the kwarg makes sense at all?

**Comment by jklymak** [comment]
OK, duh, I added `zero_formats`.  Sorry, should have looked at the code.  This is a definite bug.  I'll fix ASAP.

**Comment by jklymak** [comment]
Thanks for the report, and sorry about my confusion above.  This should be fixed in 3.1.2, in the meantime, your workaround is correct.
